singularity-eos/eos/modifiers/tilted_maxwell_eos.hpp

```diff
@@ -0,0 +1,314 @@
+                               pin, 0.8 * rho_p_max, 
+                               eos_rho_min, rho_p_max,
+                               tol, tol,
+                               rho_l, counts_l);
+    auto stat_h = regula_falsi([this, temp](Real rho){return this->t_.PressureFromDensityTemperature(rho, temp);}, 
+                               pin, 1.1 * rho_p_min, 
+                               rho_p_min, eos_rho_max,
+                               tol, tol,
+                               rho_h, counts_h);
+    // Integrate Maxwell construction integral (i.e. the chemical potential difference) 
+    Real maxwell_int = math_utils::GLQuad32([this, temp, pin](Real rho){return (this->t_.PressureFromDensityTemperature(rho, temp) - pin)/(rho*rho);}, rho_l, rho_h);
+    return std::make_tuple(maxwell_int, rho_l, rho_h);
+  }
+  
+  auto GetMaxwellConstruction(Real temp) const { 
+    constexpr double eps = std::sqrt(std::numeric_limits<Real>::epsilon());
+    
+    // Minimize dPdrho to split isotherm into two regions, if minimum is greater than zero 
+    // there is no spinodal region
+    auto dpdrho = [this, eps, temp](double rho) { 
+      return (this->t_.PressureFromDensityTemperature(rho * (1.0 + eps), temp) 
+      - this->t_.PressureFromDensityTemperature(rho * (1.0 - eps), temp)) / (2 * eps * rho);
+    };
+    auto d2pdrho2 = [this, eps, temp](double rho) { 
+      return (this->t_.PressureFromDensityTemperature(rho * (1.0 + eps), temp) 
+            - 2*this->t_.PressureFromDensityTemperature(rho, temp)
+            + this->t_.PressureFromDensityTemperature(rho * (1.0 - eps), temp)) / (eps * eps * rho * rho);
+    };
+    Real rho_dpdrho_min = 0.0; 
+    RootCounts counts;
+    auto stat = regula_falsi(d2pdrho2, 
+                             0.0, 1.0, 
+                             0.1, 3.0,
+                             tol, tol,
+                             rho_dpdrho_min, counts);
+    if (dpdrho(rho_dpdrho_min) >= 0.0) return std::make_tuple(0.0, 0.0, 0.0, false); 
+
+    // Find spinodal region by finding maximum pressure below and minimum pressure above rho_dpdrho_min
+    Real rho_p_max = 0.0; 
+    Real rho_p_min = 0.0;
+    stat = regula_falsi(dpdrho, 
+                        0.0, 0.8 * rho_dpdrho_min, 
+                        eos_rho_min, rho_dpdrho_min, 
+                        tol, tol,
+                        rho_p_max, counts);
+    stat = regula_falsi(dpdrho, 
+                        0.0, 1.2 * rho_dpdrho_min, 
+                        rho_dpdrho_min, eos_rho_max, 
+                        tol, tol,
+                        rho_p_min, counts);
+    Real p_max = std::min(t_.PressureFromDensityTemperature(rho_p_max, temp), 
+                          t_.PressureFromDensityTemperature(eos_rho_max, temp));
+    Real p_min = std::max(t_.PressureFromDensityTemperature(rho_p_min, temp),
+                          t_.PressureFromDensityTemperature(eos_rho_min, temp));
```

Comment:
this is almost identical code to in the test. Is there some way to factor out the logic to find for the edges of the spinodal region so this code doesn't need to be duplicated?

Reply:
It is close, but different. It is finding the densities corresponding to the local minimum and maximum of the pressure (i.e. the boundaries of the spinodal region), while the code in `TestMaxwellConditions` is finding a densities in a particular region that correspond to a given pressure. I am not sure that there is enough code duplication that we can factor it out.